### PR TITLE
DNM: Using openat/fstatat/linkat/unlinkat in FileStore.

### PR DESCRIPTION
**Pull request by majianpeng**
Idea from the newstore.

### CODE REVIEW

**Comment by majianpeng** [comment]
@liewegas . Sorry for my careless. I push old code. Now the correctly code. Please review again! Thank!

**Comment by athanatos** [comment]
So, at first glance, I think we definitely shouldn't mix relative and direct paths using a bool.  Is there any reason to ever consider a direct path?  That just seems like a rich source of bugs.

File: src/os/CollectionIndex.h [hunk 1]
```diff
@@ -192,6 +194,20 @@ class CollectionIndex {
       uint64_t expected_num_objs  ///< [in] expected number of objects this collection has
       ) { assert(0); return 0; }
 
+  /*
+   * Return the file descriptor for pg dir. It only use for filestore.
```
Comment (athanatos) [inline src/os/CollectionIndex.h]: CollectionIndex itself is only used for filestore...

**Comment by athanatos** [comment]
What testing have you done on this branch?

**Comment by athanatos** [comment]
The next step is to continue discussing the above questions.  After that, we'll transition to needs-qa at which point you will need to supply a pulpito link to a completed rados suite with explanations for failures and seperately some evidence of performance improvement.  After that, we'll do a review and possibly redo testing.  Once we're happy with the testing and review, we'll add it to the integration branch for final testing and merge.

Note, we are reluctant to make any serious changes to filestore without extensive testing and evidence of significant performance improvement.

**Comment by majianpeng** [comment]
@athanatos . In fact, for IndexedPath, we don't need direct path(full path). It only need relative path. So i modify code

**Comment by dachary** [comment]
@majianpeng please ignore the bot false negative (http://tracker.ceph.com/issues/13986). You can re-schedule a run by rebasing & repushing.

**Comment by liewegas** [comment]
This is too complicated a change for FileStore.

**Pull request closed without merging**